### PR TITLE
Add CI and Windows support for cl.exe

### PR DESCRIPTION
**Pull request by JoeyGrajciar**
I tried #347 and changes were looking fine(`cd` does not like `;`), but since the branch was too old I decided to do it this way.

Changes are working tested on windows machine. Also to make this work `nvim` has to be started from `Developer Command Prompt` (`cl.exe` has some environmental needs)

### CODE REVIEW

**Comment by theHamsta** [comment]
Just asking: is the a possibility that we can test this functionality on CI? 

I've started to add Windows to the CI pipelines (but with clang). It would be interesting for us to see whether the parsers compile fine on `cl.exe` or with `llvm` on Windows.

https://github.com/nvim-treesitter/nvim-treesitter/pull/417/files

File: lua/nvim-treesitter/shell_command_selectors.lua
Comment (theHamsta) [inline lua/nvim-treesitter/shell_command_selectors.lua:186]: is there maybe a more pregnant same for this? `select_push_dir_command`/``select_pop_dir_command`

File: lua/nvim-treesitter/install.lua
Comment (theHamsta) [inline lua/nvim-treesitter/install.lua:113]: It would be probably better to just have one function to set the temporary directory that can be called at the end of this function to surround the actual command.

File: lua/nvim-treesitter/shell_command_selectors.lua
Comment (theHamsta) [inline lua/nvim-treesitter/shell_command_selectors.lua:188]: Will this always work on win32? Even when using bash?

File: lua/nvim-treesitter/install.lua
Comment (JoeyGrajciar) [reply lua/nvim-treesitter/install.lua:113]: yeah you are right :) it will be nicer :)

File: lua/nvim-treesitter/shell_command_selectors.lua
Comment (JoeyGrajciar) [reply lua/nvim-treesitter/shell_command_selectors.lua:188]: hmm I guess not. this one is tricky

**Comment by JoeyGrajciar** [comment]
CI should be possible I guess :) should I add it in this PR?

File: lua/nvim-treesitter/shell_command_selectors.lua
Comment (theHamsta) [reply lua/nvim-treesitter/shell_command_selectors.lua:188]: Maybe Developer Command Prompt sets some env variables? Does something like `$SHELL` exists on Windows.

**Comment by theHamsta** [comment]
> CI should be possible I guess :) should I add it in this PR?

If have time for it you can include it in this PR. But CI usually involves more efforts than one might think beforehand. If you think it will be too much work we can also include it in a follow-up PR. Windows CI would help us (all maintainers are Linux or sometimes Mac users) and Windows users a lot since sometimes Windows users submit patches but they will not work when other users use another shell/compiler combination.

File: lua/nvim-treesitter/shell_command_selectors.lua
Comment (JoeyGrajciar) [reply lua/nvim-treesitter/shell_command_selectors.lua]: hmm so `pushd` and `popd` do work locally on Mac, but not on CI in ubuntu.

Comment (JoeyGrajciar) [reply lua/nvim-treesitter/shell_command_selectors.lua]: checked those commands in Git Bash and they should work. was able to compile `cpp` with clang.

**Comment by JoeyGrajciar** [comment]
> If have time for it you can include it in this PR.

Will check this later today.

**Comment by theHamsta** [comment]
image
Classic CI work :smile: . So glad that I'm not an dev opt engineer.

File: .github/workflows/test.yml
Comment (theHamsta) [inline .github/workflows/test.yml:19]: Does `:TSSync all` work? Then `sleep` would not be necessary.

File: .github/workflows/test.yml
Comment (theHamsta) [inline .github/workflows/test.yml]: maybe it requires a `refreshenv`?

File: .github/workflows/test.yml
Comment (JoeyGrajciar) [reply .github/workflows/test.yml]: apparently it works :)

File: .github/workflows/test.yml
Comment (JoeyGrajciar) [reply .github/workflows/test.yml]: currently went with complete path to `nvim.exe`, but seems there are some failures for some languages and also `gcc` appeared in the wild :)

Comment (JoeyGrajciar) [reply .github/workflows/test.yml]: okay so issue was with those "special" ones with location specified in parser config

**Comment by theHamsta** [comment]
Looks nice! Do the uploaded parsers from CI work on you machine? In the past some users said that the shared libraries would not be accepted by their OS when loading.

**Comment by JoeyGrajciar** [comment]
> Do the uploaded parsers from CI work on you machine?

what uploaded parsers?

File: lua/nvim-treesitter/shell_command_selectors.lua
Comment (theHamsta) [inline lua/nvim-treesitter/shell_command_selectors.lua:188]: We did not solve that issue yet, whether to use pushd or cd on Windows, didn't we?

Comment (JoeyGrajciar) [reply lua/nvim-treesitter/shell_command_selectors.lua:188]: `cd %s;\n` does not work on Windows :) It does not like that `;`

Comment (JoeyGrajciar) [reply lua/nvim-treesitter/shell_command_selectors.lua:188]: I mean now we can test it quite easily in CI :)

**Comment by JoeyGrajciar** [comment]
Should I try to add also macOS CI?

**Comment by theHamsta** [comment]
This will upload the compiled parsers when CI finishes
```
      - uses: actions/upload-artifact@v2
        with:
          name: parsers
          path: parser/*
```

**Comment by theHamsta** [comment]
I think MacOS is not so different from Linux. So there is no urgent need to add it but it won't harm.

**Comment by JoeyGrajciar** [comment]
Autosquash history, renamed CI commit (its not windows only anymore) :)
Will try to test those artifacts today.

File: lua/nvim-treesitter/shell_command_selectors.lua
Comment (theHamsta) [reply lua/nvim-treesitter/shell_command_selectors.lua:188]: But bash on windows should behave like bash on Linux?

Comment (JoeyGrajciar) [reply lua/nvim-treesitter/shell_command_selectors.lua:188]: you mean running this installation command from `bash` in CI?

Comment (theHamsta) [reply lua/nvim-treesitter/shell_command_selectors.lua:188]: I mean when users on windows use bash instead of powershell. Will they have a problem? I can't really tell what's the common eniviroment for running nvim on windows (probably neovim-qt) but users seem to have managed to install parsers also with cd;

Comment (JoeyGrajciar) [reply lua/nvim-treesitter/shell_command_selectors.lua:188]: I mean probably most of the shell commands which do check `win32` will not work on bash.

**Comment by JoeyGrajciar** [comment]
> In the past some users said that the shared libraries would not be accepted by their OS when loading.

gcc on windows for some reason generates some libraries with dependencies on `libstdc++-6.dll` and `libgcc_s_seh-1.dll`
other issue might be trying to load x86_64 on x86 system and vice versa (same for arm)

**Comment by theHamsta** [comment]
> gcc on windows for some reason generates some libraries with dependencies on libstdc++-6.dll and libgcc_s_seh-1.dll
other issue might be trying to load x86_64 on x86 system and vice versa (same for arm)

@JoeyGrajciar  I don't think that this was the reason. The binaries depend on libstdc++ due to the linker flag `-lstdc++`. We need this flag because some scanners use C++ for `scanner.cc`. `cl.exe` made historically never a real difference between C and C++

**Review by theHamsta — APPROVED** [review]

**Comment by JoeyGrajciar** [comment]
> The binaries depend on libstdc++ due to the linker flag -lstdc++. We need this flag because some scanners use C++ for scanner.cc. cl.exe made historically never a real difference between C and C++

I meant that some parsers built on windows-latest with gcc are not getting loaded on windows due to those dependencies.

**Comment by theHamsta** [comment]
Probably we can update the Windows wiki entry a bit after merging this. If someone cannot get the parsers compiled at least they can download them from CI.

**Comment by JoeyGrajciar** [comment]
Okay so after some playing with that compilation locally I found the issue :)
gcc that is available during compilation (PATH) on `windows-latest` is from Mingw 8.1.0 and it creates these dependencies
image

After upgrading mentioned Mingw to latest 10.2.0 it looks like this for the same file
image

information about them is the key
```
gcc -v
Using built-in specs.
COLLECT_GCC=C:\ProgramData\Chocolatey\lib\mingw\tools\install\mingw64\bin\gcc.exe
COLLECT_LTO_WRAPPER=C:/ProgramData/Chocolatey/lib/mingw/tools/install/mingw64/bin/../libexec/gcc/x86_64-w64-mingw32/8.1.0/lto-wrapper.exe
Target: x86_64-w64-mingw32
Configured with: ...
Thread model: posix
gcc version 8.1.0 (x86_64-posix-seh-rev0, Built by MinGW-W64 project) 
```

vs

```
gcc -v
Using built-in specs.
COLLECT_GCC=C:\ProgramData\chocolatey\lib\mingw\tools\install\mingw64\bin\gcc.exe
COLLECT_LTO_WRAPPER=c:/programdata/chocolatey/lib/mingw/tools/install/mingw64/bin/../libexec/gcc/x86_64-w64-mingw32/10.2.0/lto-wrapper.exe
Target: x86_64-w64-mingw32
Configured with: ...
Thread model: win32
Supported LTO compression algorithms: zlib
gcc version 10.2.0 (GCC)
```

AFAIK they are not going to upgrade this Mingw to latest version. Instead there is Msys2 installed which also contains gcc.
Will investigate option to use that one instead or we could try to install latest Mingw as part of setup.

**Comment by theHamsta** [comment]
@JoeyGrajciar I think it's ok when artifacts from cl.exe are working or gcc/clang builds locally

**Comment by JoeyGrajciar** [comment]
@theHamsta do you mean that I should disable creation of artifact for gcc on `windows-latest`?

**Comment by theHamsta** [comment]
@JoeyGrajciar you could, but I would just link to the CL.exe binaries in the wiki. I'd say everyone using the artifacts is at their own risk.

**Comment by theHamsta** [comment]
I fear that the MacOS brew method might be broken because tree-sitter has been updated to 0.19.0. Could merge since the button is still green 😄

**Comment by JoeyGrajciar** [comment]
> I fear that the MacOS brew method might be broken because tree-sitter has been updated to 0.19.0. Could merge since the button is still green

for now it is good 😄 rebased on top of master and looks fine even though it was downloading 0.18.2

**Comment by theHamsta** [comment]
@vigoux @kyazdani42 any comments to the new cross-platform CI? It's a lot of jobs but now every build configuration is tested. Great work to have our plugin work on all platforms with all compilers!

**Comment by JoeyGrajciar** [comment]
> It's a lot of jobs but now every build configuration is tested.

That is very bold statement 😁it is probably most used common combinations 🙂

**Review by kyazdani42 — APPROVED** [review]
good job indeed

**Comment by theHamsta** [comment]
@JoeyGrajciar can you update the wiki with whatever magic that was necessary to make this work and how to download parsers from CI?

**Comment by JoeyGrajciar** [comment]
@theHamsta wiki updated.

**Pull request merged**
